### ISSUE TITLE
regression with getfield access on global objects

### ISSUE THREAD

**Issue by rfourquet** [JuliaLang/julia#50870]
Compared to v1.8:
* v1.9.2 has a regression where `a.xx` allocates (whereas `getfield(a, :xx)` doesn't) 
* v1.10.0-beta1 has a regression where both `a[1].xx` and `getfield(a[1], :xx)` allocate when `a` is a a tuple

The following script should show the problem:
```julia
using BenchmarkTools

struct XX
    xx::Vector{Int}
end

const XXs = XX(Int[])

get1() = length(getfield(XXs, :xx))
get2() = length(XXs.xx)

const YYs = (XX(Int[]),)

get3() = length(getfield(YYs[1], :xx))
get4() = length(YYs[1].xx)

@btime get1()
@btime get2()
@btime get3()
@btime get4()
nothing
```
Then:
```
julia> include("tmp/julia/getfield.jl") # v1.8.5
  1.923 ns (0 allocations: 0 bytes)
  1.924 ns (0 allocations: 0 bytes)
  1.923 ns (0 allocations: 0 bytes)
  1.923 ns (0 allocations: 0 bytes)

julia> include("tmp/julia/getfield.jl") # v1.9.2
  1.923 ns (0 allocations: 0 bytes)
  12.877 ns (1 allocation: 16 bytes)
  2.134 ns (0 allocations: 0 bytes)
  12.927 ns (1 allocation: 16 bytes)

julia> include("tmp/julia/getfield.jl") # v1.10.0-beta1
  1.953 ns (0 allocations: 0 bytes)
  2.815 ns (0 allocations: 0 bytes)
  15.269 ns (1 allocation: 16 bytes)
  17.970 ns (1 allocation: 16 bytes)
```

**Comment by gbaraldi** [comment]
This is fixed on master and the fix is getting backported to both 1.9 and 1.10 ;)
https://github.com/JuliaLang/julia/pull/50523

**Comment by rfourquet** [comment]
> This is fixed on master and the fix is getting backported to both 1.9 and 1.10 ;)

Ah cool! I was compiling master to try it out but then forgot about it ;-)
